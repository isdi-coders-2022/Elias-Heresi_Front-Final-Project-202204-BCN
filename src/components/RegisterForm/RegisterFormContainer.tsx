import styled from "styled-components";

export const RegisterFormContainer = styled.div`
  border: 2px grey solid;
  padding: 20px 30px;
  border-radius: 25px;
  background-color: rgb(151, 227, 168);
  width: 90%;
  min-width: 300px;
  max-width: 400px;

  form {
    div {
      font-weight: bold;
      font-size: 24px;
    }
    section {
      font-size: 24px;
    }
  }
`;
