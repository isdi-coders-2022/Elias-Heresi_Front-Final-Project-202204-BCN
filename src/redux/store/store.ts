import { configureStore } from "@reduxjs/toolkit";
import userReducer from "../features/userSlice";
import uiReducer from "../features/uiSlice";
import diaryReducer from "../features/diarySlice";

const store = configureStore({
  reducer: {
    user: userReducer,
    ui: uiReducer,
    diary: diaryReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
