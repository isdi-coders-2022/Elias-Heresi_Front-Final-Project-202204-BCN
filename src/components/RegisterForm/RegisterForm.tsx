import { ChangeEvent, useState, FormEvent, useEffect } from "react";
import { Button, Form } from "react-bootstrap";
import { RegisterFormContainer } from "./RegisterFormContainer";
import { RegisterInformation } from "../../redux/interfaces/userInterfaces";
import { registerUserThunk } from "../../redux/thunks/thunks";
import { useAppDispatch } from "../../redux/store/hooks";

const RegisterForm = (): JSX.Element => {
  const formInitialState = {
    name: "",
    surname: "",
    email: "",
    username: "",
    password: "",
  } as RegisterInformation;

  const [formData, setFormData] = useState(formInitialState);
  const [buttonDisabled, setButtonDisabled] = useState(true);

  const dispatch = useAppDispatch();

  const changeData = (event: ChangeEvent<HTMLInputElement>): void => {
    setFormData({ ...formData, [event.target.id]: event.target.value });
  };

  const resetForm = (): void => {
    setFormData(formInitialState);
  };

  const submitRegister = (event: FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    dispatch(registerUserThunk(formData));
    resetForm();
  };

  useEffect(() => {
    if (
      formData.name &&
      formData.email &&
      formData.surname &&
      formData.password &&
      formData.username
    ) {
      setButtonDisabled(false);
      return;
    }
    setButtonDisabled(true);
  }, [formData]);

  return (
    <RegisterFormContainer>
      <Form autoComplete="off" onSubmit={submitRegister} noValidate>
        <Form.Group className="mb-3">
          <Form.Label>Name</Form.Label>
          <Form.Control
            id="name"
            type="text"
            placeholder="Enter first name"
            value={formData.name}
            onChange={changeData}
          />
        </Form.Group>

        <Form.Group className="mb-3">
          <Form.Label>Surname</Form.Label>
          <Form.Control
            id="surname"
            type="text"
            placeholder="Enter last name"
            value={formData.surname}
            onChange={changeData}
          />
        </Form.Group>

        <Form.Group className="mb-3">
          <Form.Label>Email address</Form.Label>
          <Form.Control
            id="email"
            type="email"
            placeholder="Enter email"
            value={formData.email}
            onChange={changeData}
          />
        </Form.Group>

        <Form.Group className="mb-3">
          <Form.Label>Username</Form.Label>
          <Form.Control
            id="username"
            type="text"
            placeholder="Choose your username"
            value={formData.username}
            onChange={changeData}
          />
        </Form.Group>

        <Form.Group className="mb-3">
          <Form.Label>Password</Form.Label>
          <Form.Control
            id="password"
            type="password"
            placeholder="Choose your password"
            value={formData.password}
            onChange={changeData}
          />
        </Form.Group>

        <section className="container text-center">
          <Button variant="primary" type="submit" disabled={buttonDisabled}>
            Submit
          </Button>

          <p>Already a member?</p>

          <Button variant="secondary" type="button">
            Login
          </Button>
        </section>
      </Form>
    </RegisterFormContainer>
  );
};

export default RegisterForm;
