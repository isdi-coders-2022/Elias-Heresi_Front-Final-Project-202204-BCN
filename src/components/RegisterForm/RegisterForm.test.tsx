import { Provider } from "react-redux";
import RegisterForm from "./RegisterForm";
import store from "../../redux/store/store";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TestRenderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

const mockUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUseNavigate,
}));

describe("Given the RegisterForm component", () => {
  describe("When invoked", () => {
    test("Then 2 buttons and 4 input boxes will be rendered", () => {
      const expectedButtons = 2;
      const expectedInputBoxes = 4;

      render(
        <BrowserRouter>
          <Provider store={store}>
            <RegisterForm />
          </Provider>
        </BrowserRouter>
      );

      const searchedButtons = screen.getAllByRole("button");
      const searchedInputBoxes = screen.getAllByRole("textbox");

      expect(searchedButtons).toHaveLength(expectedButtons);
      expect(searchedInputBoxes).toHaveLength(expectedInputBoxes);
    });
  });
  describe("When the register button is clicked on", () => {
    test("Then the user will be redirected to another page", () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <RegisterForm />
          </BrowserRouter>
        </Provider>
      );

      const registerButton = screen.getByRole("button", { name: "Login" });
      userEvent.click(registerButton);

      expect(mockUseNavigate).toHaveBeenCalled();
    });
  });
});
describe("When invoked, and the submit button is enabled and clicked on", () => {
  test("Then an action will be dispatched", () => {
    render(
      <BrowserRouter>
        <Provider store={store}>
          <RegisterForm />
        </Provider>
      </BrowserRouter>
    );

    const nameInputText = screen.getByLabelText("Name");
    const surnameInputText = screen.getByLabelText("Surname");
    const emailInputText = screen.getByLabelText("Email address");
    const usernameInputText = screen.getByLabelText("Username");
    const passwordInputText = screen.getByLabelText("Password");

    userEvent.type(nameInputText, "text");
    userEvent.type(surnameInputText, "text");
    userEvent.type(emailInputText, "text");
    userEvent.type(usernameInputText, "text");
    userEvent.type(passwordInputText, "text");

    const submitButton = screen.getByRole("button", { name: "Submit" });
    userEvent.click(submitButton);

    expect(mockDispatch).toHaveBeenCalled();
  });

  test("Then it should always match this snapshot", () => {
    const testedLoginPage = TestRenderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <RegisterForm />
        </BrowserRouter>
      </Provider>
    );

    expect(testedLoginPage).toMatchSnapshot();
  });
});
