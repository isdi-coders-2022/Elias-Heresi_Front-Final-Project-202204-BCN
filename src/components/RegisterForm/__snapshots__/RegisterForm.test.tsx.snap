// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When invoked, and the submit button is enabled and clicked on Then it should always match this snapshot 1`] = `
<div
  className="sc-bczRLJ fUXCFd"
>
  <form
    autoComplete="off"
    className=""
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="name"
      >
        Name
      </label>
      <input
        className="form-control"
        id="name"
        onChange={[Function]}
        placeholder="Enter first name"
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="surname"
      >
        Surname
      </label>
      <input
        className="form-control"
        id="surname"
        onChange={[Function]}
        placeholder="Enter last name"
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="email"
      >
        Email address
      </label>
      <input
        className="form-control"
        id="email"
        onChange={[Function]}
        placeholder="Enter email"
        type="email"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="username"
      >
        Username
      </label>
      <input
        className="form-control"
        id="username"
        onChange={[Function]}
        placeholder="Choose your username"
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="form-control"
        id="password"
        onChange={[Function]}
        placeholder="Choose your password"
        type="password"
        value=""
      />
    </div>
    <section
      className="container text-center"
    >
      <button
        className="btn btn-primary"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
      <p>
        Already a member?
      </p>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Login
      </button>
    </section>
  </form>
</div>
`;
